[3.6] bpo-33312: Fix clang ubsan out of bounds warnings in dict. (GH-6537)

miss-islington opened the pull request:
Fix clang ubsan (undefined behavior sanitizer) warnings in dictobject.c by
adjusting how the internal struct _dictkeysobject shared keys structure is
declared.

This remains ABI compatible.  We get rid of the union at the end of the
struct being used for conveinence to avoid typecasting in favor of char[]
variable length array at the end of a struct. This is known to clang to be
used for variable sized objects and will not cause an undefined behavior
problem.  Similarly, char arrays do not have strict aliasing undefined
behavior when cast.

PEP-007 does not currently list variable length arrays (VLAs) as allowed
in our subset of C99.  If this turns out to be a problem, the fix to this is
to change the char `dk_indices[]` into `dk_indices[1]` and restore the
three size computation subtractions this change removes:
  `- Py_MEMBER_SIZE(PyDictKeysObject, dk_indices)`

If this works as is I'll make a separate PR to update PEP-007.
(cherry picked from commit 397f1b28c4a12e3b3ed59a89599eabc457412649)


Co-authored-by: Gregory P. Smith <greg@krypto.org>

<!-- issue-number: [bpo-33312](https://www.bugs.python.org/issue33312) -->
https://bugs.python.org/issue33312
<!-- /issue-number -->

gpshead committed:
bpo-33312: Fix clang ubsan out of bounds warnings in dict. (GH-6537)

Fix clang ubsan (undefined behavior sanitizer) warnings in dictobject.c by
adjusting how the internal struct _dictkeysobject shared keys structure is
declared.

This remains ABI compatible.  We get rid of the union at the end of the
struct being used for conveinence to avoid typecasting in favor of char[]
variable length array at the end of a struct. This is known to clang to be
used for variable sized objects and will not cause an undefined behavior
problem.  Similarly, char arrays do not have strict aliasing undefined
behavior when cast.

PEP-007 does not currently list variable length arrays (VLAs) as allowed
in our subset of C99.  If this turns out to be a problem, the fix to this is
to change the char `dk_indices[]` into `dk_indices[1]` and restore the
three size computation subtractions this change removes:
  `- Py_MEMBER_SIZE(PyDictKeysObject, dk_indices)`

If this works as is I'll make a separate PR to update PEP-007.
(cherry picked from commit 397f1b28c4a12e3b3ed59a89599eabc457412649)

Co-authored-by: Gregory P. Smith <greg@krypto.org>

gpshead reviewed (changes requested):
I need to merge https://github.com/python/cpython/pull/6549/files into this PR before it goes in.

gpshead committed:
Fix Tools/gdb/libpython.py

Backport https://github.com/python/cpython/commit/11659d00b9185c8f02ea6b642fa475a80e21f1a9
into this change instead of leaving it a separate followup change.

gpshead reviewed (changes requested):
even after tests pass, i'm going to let this sit for a while watching for any unexpected fallout on the master and 3.7 sides.
